slam/src/slam/fast_slam2/FastSLAM2.py

```diff
@@ -0,0 +1,124 @@
+from collections import namedtuple
+from slam.fast_slam2.Models import Meas, FastSLAM2Parameters, LandmarkConstants
+
+class FastSLAM2():
+  def __init__(self, robot_physics, parameters: FastSLAM2Parameters, random=np.random.default_rng()):
+
+    self.num_particles = parameters.num_particles
+    self.particles = []
+
+    self._meas_ls: list[Meas] = []
+    self._prev_t = 0
+    self._random = random
+    self._particles_params = {
+      'random': self._random,
+      'are_landmarks_fixed': parameters.are_landmarks_fixed,
+      'initial_landmarks': parameters.initial_landmarks,
+      'initial_pose': parameters.initial_pose,
+      'initial_pose_cov': parameters.initial_pose_cov,
+      'landmark_constants': parameters.landmark_constants,
+      'verbose': parameters.verbose
+    }
+    self._localization_only = parameters.localization_only
+    self._verbose = parameters.verbose
+    self._robot_physics = robot_physics
+    self._fast_slam_version = parameters.fast_slam_version
+
+    self._log(parameters)
+    self._create_particles(self.num_particles)
+
+  def _create_particles(self, n):
+    for i in range(n):
+      if self._fast_slam_version == 1:
+        self.particles.append(FastSLAM1Particle(robot_physics = self._robot_physics, particle_id=i, **self._particles_params))
+      else:
+        self.particles.append(FastSLAM2Particle(robot_physics = self._robot_physics, particle_id=i, **self._particles_params))
+    self._log('info', 'fast slam 2 initial pose', self._particles_params['initial_pose'])
+
+  def add_control(self, control, time):
+    self._update_meas()
+    if len(self.particles) > 1 and len(self._meas_ls) > 0: # only resample if there are more than one particle and observations
+      self._resample()
+    self._log('control',control)
+    self._update_motion(control, time)
+    self._meas_ls = []
+
+  def add_measurement(self, meas: Meas):
+    self._meas_ls.append(meas)
+
+  def _log(self, *msg):
+    if self._verbose >= 1:
+      print('FastSLAM2:', *msg)
+
+  def _update_motion(self, control, time): #propagate motion
+    dt = max(time - self._prev_t, 0.000001)
+    if (time - self._prev_t == 0): self._log("warning:","dt is 0")
+    for particle in self.particles:
+      particle.update_motion(control, dt)
+    self._prev_t = time
+
+  def _update_meas(self):
+    for particle in self.particles:
+      if self._localization_only:
+        particle.update_meas_localization_only(self._meas_ls)
+      else:
+        particle.update_meas(self._meas_ls)
+
+  def _resample(self):    #  Collect weight
+    self.weights = np.array([particle.weight for particle in self.particles])
+    self.weights = self.weights / np.sum(self.weights)
+    for idx, particle in enumerate(self.particles):
+      particle.accumulated_weight += self.weights[idx]
+
+    try: # resampling
+      new_particle_idx_ls = self._random.choice(list(range(self.num_particles)), size=self.num_particles, replace=True, p=self.weights)
+    except:
```

Comment:
Consider only catching the actual error rather than all exceptions. Not a big deal, but good practice.